From 11840788815d6d0459ae6911b8f8ba5d11928a87 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Tue, 24 Dec 2024 16:45:56 +0800
Subject: [PATCH 13/23] Revert "fs: Allow fine-grained control of folio sizes"

hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBDFAV

----------------------------------------------------------------------

This reverts commit e6ebbdef6b5db01872e4c5866d333cb2f923be57.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 include/linux/pagemap.h | 90 ++++++-----------------------------------
 mm/readahead.c          |  4 +-
 2 files changed, 15 insertions(+), 79 deletions(-)

diff --git a/include/linux/pagemap.h b/include/linux/pagemap.h
index e44e377661f2..429627abfef4 100644
--- a/include/linux/pagemap.h
+++ b/include/linux/pagemap.h
@@ -203,21 +203,12 @@ enum mapping_flags {
 	AS_EXITING	= 4, 	/* final truncate in progress */
 	/* writeback related tags are not used */
 	AS_NO_WRITEBACK_TAGS = 5,
-	AS_RELEASE_ALWAYS = 6,	/* Call ->release_folio(), even if no private data */
-	AS_STABLE_WRITES = 7,	/* must wait for writeback before modifying
+	AS_LARGE_FOLIO_SUPPORT = 6,
+	AS_RELEASE_ALWAYS,	/* Call ->release_folio(), even if no private data */
+	AS_STABLE_WRITES,	/* must wait for writeback before modifying
 				   folio contents */
-	AS_INACCESSIBLE = 8,	/* Do not attempt direct R/W access to the mapping */
-	/* Bits 16-25 are used for FOLIO_ORDER */
-	AS_FOLIO_ORDER_BITS = 5,
-	AS_FOLIO_ORDER_MIN = 16,
-	AS_FOLIO_ORDER_MAX = AS_FOLIO_ORDER_MIN + AS_FOLIO_ORDER_BITS,
 };
 
-#define AS_FOLIO_ORDER_BITS_MASK ((1u << AS_FOLIO_ORDER_BITS) - 1)
-#define AS_FOLIO_ORDER_MIN_MASK (AS_FOLIO_ORDER_BITS_MASK << AS_FOLIO_ORDER_MIN)
-#define AS_FOLIO_ORDER_MAX_MASK (AS_FOLIO_ORDER_BITS_MASK << AS_FOLIO_ORDER_MAX)
-#define AS_FOLIO_ORDER_MASK (AS_FOLIO_ORDER_MIN_MASK | AS_FOLIO_ORDER_MAX_MASK)
-
 /**
  * mapping_set_error - record a writeback error in the address_space
  * @mapping: the mapping in which an error should be set
@@ -357,51 +348,9 @@ static inline void mapping_set_gfp_mask(struct address_space *m, gfp_t mask)
 #define MAX_XAS_ORDER		(XA_CHUNK_SHIFT * 2 - 1)
 #define MAX_PAGECACHE_ORDER	min(MAX_XAS_ORDER, PREFERRED_MAX_PAGECACHE_ORDER)
 
-/*
- * mapping_set_folio_order_range() - Set the orders supported by a file.
- * @mapping: The address space of the file.
- * @min: Minimum folio order (between 0-MAX_PAGECACHE_ORDER inclusive).
- * @max: Maximum folio order (between @min-MAX_PAGECACHE_ORDER inclusive).
- *
- * The filesystem should call this function in its inode constructor to
- * indicate which base size (min) and maximum size (max) of folio the VFS
- * can use to cache the contents of the file.  This should only be used
- * if the filesystem needs special handling of folio sizes (ie there is
- * something the core cannot know).
- * Do not tune it based on, eg, i_size.
- *
- * Context: This should not be called while the inode is active as it
- * is non-atomic.
- */
-static inline void mapping_set_folio_order_range(struct address_space *mapping,
-						 unsigned int min,
-						 unsigned int max)
-{
-	if (!IS_ENABLED(CONFIG_TRANSPARENT_HUGEPAGE))
-		return;
-
-	if (min > MAX_PAGECACHE_ORDER)
-		min = MAX_PAGECACHE_ORDER;
-
-	if (max > MAX_PAGECACHE_ORDER)
-		max = MAX_PAGECACHE_ORDER;
-
-	if (max < min)
-		max = min;
-
-	mapping->flags = (mapping->flags & ~AS_FOLIO_ORDER_MASK) |
-		(min << AS_FOLIO_ORDER_MIN) | (max << AS_FOLIO_ORDER_MAX);
-}
-
-static inline void mapping_set_folio_min_order(struct address_space *mapping,
-					       unsigned int min)
-{
-	mapping_set_folio_order_range(mapping, min, MAX_PAGECACHE_ORDER);
-}
-
 /**
  * mapping_set_large_folios() - Indicate the file supports large folios.
- * @mapping: The address space of the file.
+ * @mapping: The file.
  *
  * The filesystem should call this function in its inode constructor to
  * indicate that the VFS can use large folios to cache the contents of
@@ -412,23 +361,7 @@ static inline void mapping_set_folio_min_order(struct address_space *mapping,
  */
 static inline void mapping_set_large_folios(struct address_space *mapping)
 {
-	mapping_set_folio_order_range(mapping, 0, MAX_PAGECACHE_ORDER);
-}
-
-static inline unsigned int
-mapping_max_folio_order(const struct address_space *mapping)
-{
-	if (!IS_ENABLED(CONFIG_TRANSPARENT_HUGEPAGE))
-		return 0;
-	return (mapping->flags & AS_FOLIO_ORDER_MAX_MASK) >> AS_FOLIO_ORDER_MAX;
-}
-
-static inline unsigned int
-mapping_min_folio_order(const struct address_space *mapping)
-{
-	if (!IS_ENABLED(CONFIG_TRANSPARENT_HUGEPAGE))
-		return 0;
-	return (mapping->flags & AS_FOLIO_ORDER_MIN_MASK) >> AS_FOLIO_ORDER_MIN;
+	__set_bit(AS_LARGE_FOLIO_SUPPORT, &mapping->flags);
 }
 
 /**
@@ -442,7 +375,7 @@ mapping_min_folio_order(const struct address_space *mapping)
 static inline void mapping_clear_large_folios(struct address_space *mapping)
 {
 	WARN_ON_ONCE(!rwsem_is_locked(&mapping->invalidate_lock));
-	mapping_set_folio_order_range(mapping, 0, 0);
+	__clear_bit(AS_LARGE_FOLIO_SUPPORT, &mapping->flags);
 }
 
 /*
@@ -451,17 +384,20 @@ static inline void mapping_clear_large_folios(struct address_space *mapping)
  */
 static inline bool mapping_large_folio_support(struct address_space *mapping)
 {
-	/* AS_FOLIO_ORDER is only reasonable for pagecache folios */
+	/* AS_LARGE_FOLIO_SUPPORT is only reasonable for pagecache folios */
 	VM_WARN_ONCE((unsigned long)mapping & PAGE_MAPPING_ANON,
 			"Anonymous mapping always supports large folio");
 
-	return mapping_max_folio_order(mapping) > 0;
+	return IS_ENABLED(CONFIG_TRANSPARENT_HUGEPAGE) &&
+		test_bit(AS_LARGE_FOLIO_SUPPORT, &mapping->flags);
 }
 
 /* Return the maximum folio size for this pagecache mapping, in bytes. */
-static inline size_t mapping_max_folio_size(const struct address_space *mapping)
+static inline size_t mapping_max_folio_size(struct address_space *mapping)
 {
-	return PAGE_SIZE << mapping_max_folio_order(mapping);
+	if (mapping_large_folio_support(mapping))
+		return PAGE_SIZE << MAX_PAGECACHE_ORDER;
+	return PAGE_SIZE;
 }
 
 static inline int filemap_nr_thps(struct address_space *mapping)
diff --git a/mm/readahead.c b/mm/readahead.c
index c13c130efcca..438f142a3e74 100644
--- a/mm/readahead.c
+++ b/mm/readahead.c
@@ -513,10 +513,10 @@ void page_cache_ra_order(struct readahead_control *ractl,
 
 	limit = min(limit, index + ra->size - 1);
 
-	if (new_order < mapping_max_folio_order(mapping))
+	if (new_order < MAX_PAGECACHE_ORDER)
 		new_order += 2;
 
-	new_order = min(mapping_max_folio_order(mapping), new_order);
+	new_order = min_t(unsigned int, MAX_PAGECACHE_ORDER, new_order);
 	new_order = min_t(unsigned int, new_order, ilog2(ra->size));
 
 	/* See comment in page_cache_ra_unbounded() */
-- 
2.25.1

